### DISCUSSION TITLE
UserControl does not respond to changes in ObservableCollection at runtime

### DISCUSSION THREAD

**Discussion by mezmeri** [AvaloniaUI/Avalonia#17371] [Q&A] [answered]
Hello,

I am currently facing an issue with my UserControl not updating a ListBox when an ObservableCollection in its ViewModel is updated.

### **What I've tried**
1. I've checked the DataContext with the Debug Tool, and it is set correctly.
2. Setting the DataContext through XAML.
3. Checking that data is actually added to the ObservableCollection. It _is_ added.
4. Tried using DI because I thought that it was due to multiple instances of the ViewModel. Didn't work.

### **Behaviours**
**1. Adding data from the get-go**
The ObservableCollection is set in the ViewModel constructor and is empty at first. If I add data to it from the constructor, the data will show without any problems (both design-time and run-time).

**2. Setting IsVisible at run-time.** 
I tried binding a TextBlock.IsVisible to a boolean property inside the ViewModel, which gets set to false when clicking a button. Didn't work. However, setting the boolean inside the ViewModel to false at design/compile time, works.

**3. Using a Window instead**
If I use a Window instead of a UserControl everything works. Both the IsVisible and the ObservableCollection.

### **Important notes**
I am using a SplitView inside my MainWindow.xaml, which is then displaying the UserControl in a TransitioningContentControl. Could this be what is causing the problems?

### Code

**UploadBeatViewModel**
```cs
public partial class UploadBeatViewModel : ViewModelBase
{
    #region Repos
    private readonly BeatRepository _beatsRepository;
    #endregion

    #region Collections
    public ObservableCollection<BeatViewModel> BeatsVM { get; private set; }
    #endregion

    #region Properties
    /// <summary>
    /// <para>Determines if the user has added anything to the collection.</para>
    /// <para>Currently not working.</para>
    /// </summary>
    [ObservableProperty]
    private bool _hasUserUploaded = false;
    #endregion

    public UploadBeatViewModel()
    {
        BeatsVM = new();
        _beatsRepository = BeatRepository.Instance;
    }

#region Methods
/// <summary>
/// Creates a new Beat instance and sets the Beat.AudioFilePath to the path given by the user. The Beat instance will be added to the BeatRepo. Besides that it creates a BeatViewModel which gets added to the BeatsVM collection.
/// </summary>
/// <param name="fileToBeUploaded">The list of files to be uploaded.</param>
/// <returns>
/// A boolean which will tell if the files were successfully added to the collection.
/// </returns>
public bool CreateNewBeatAndAddAudioFile(FileInfo fileToBeUploaded)
{
    bool areFilesUploaded = false;
    Beat beatToBeCreated = new();

    beatToBeCreated.Title = "Frank Ocean Type Beat";
    beatToBeCreated.AudioFilePath = fileToBeUploaded.FullName;

    BeatViewModel beatViewModel = new(beatToBeCreated);
    BeatsVM.Add(beatViewModel);
    _beatsRepository.Add(beatToBeCreated);

    areFilesUploaded = AreFilesWithinCollectionAndRepo(beatViewModel, beatToBeCreated);
    return areFilesUploaded;
}

/// <summary>
/// Quick method to check if the files have been added to the <see cref="BeatsVM"/> and <see cref="_beatsRepository"/>.
/// </summary>
/// <param name="beatViewModel"></param>
/// <param name="beatModel"></param>
/// <returns></returns>
private bool AreFilesWithinCollectionAndRepo(BeatViewModel beatViewModel, Beat beatModel)
{
    bool isWithinBeatsVM = BeatsVM.Contains(beatViewModel);
    bool isWithinRepo = _beatsRepository.GetAll().Contains(beatModel);
    if(isWithinBeatsVM && isWithinRepo)
    {
        return true;
    }

    return false;
}
#endregion
}
```

**UploadBeatView.xaml**
```xaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="1000" d:DesignHeight="450"
			 xmlns:vm="using:BeatShare.ViewModels"
             x:Class="BeatShare.Views.UploadBeatView"
			 x:DataType="vm:UploadBeatViewModel">

	<Grid x:Name="Grid_DropGrid" RowDefinitions="40,*" Width="1000" MaxHeight="1000" DragDrop.AllowDrop="True">
		<Rectangle Grid.RowSpan="2" DragDrop.AllowDrop="True" Fill="Transparent"/>
		<TextBlock Grid.Row="0" FontSize="28" Text="Videos" HorizontalAlignment="Left" FontWeight="Bold" Margin="0,20,0,0" DragDrop.AllowDrop="True"/>

		<StackPanel x:Name="StackPanel_NoUploadsFound" IsVisible="False" Grid.Row="1" Spacing="20" VerticalAlignment="Center" DragDrop.AllowDrop="True">
			<PathIcon Data="{StaticResource search_info_regular}" Height="50" Foreground="Gray" IsVisible="{Binding !HasUserUploaded, UpdateSourceTrigger=PropertyChanged}"/>
			<TextBlock FontSize="18" TextAlignment="Center" Foreground="Gray" FontWeight="Bold" HorizontalAlignment="Center">
				No files uploaded. <LineBreak/>Drag them into this window or click the button below to start!
			</TextBlock>
			<Button Content="Upload" Theme="{StaticResource SolidButton}" Classes="Primary" Padding="30, 12"/>
		</StackPanel>
		
		<ListBox ItemsSource="{Binding BeatsVM}" Grid.Row="1" HorizontalAlignment="Center">
			<ListBox.ItemTemplate>
				<DataTemplate>
					<Border Background="Red" Height="20" Width="100">
						<StackPanel Orientation="Horizontal">
							<TextBlock Text="{Binding Title}"/>
						</StackPanel>
					</Border>
				</DataTemplate>
			</ListBox.ItemTemplate>
		</ListBox>
	</Grid>
</UserControl>
```

Code-behind:
I am rewriting the drag-drop logic, so it currently doesn't work.

```cs
public partial class UploadBeatView : UserControl
{
    private UploadBeatViewModel _viewModel;
    private bool _isDropHandlerAdded = false;
    /// <summary>
    /// Run-time constructor.
    /// </summary>
    /// <param name="service"></param>
    public UploadBeatView()
    {
        InitializeComponent();
        _viewModel = new();
        DataContext = _viewModel;
    }

    private void SetupDropHandler()
    {
        if (_isDropHandlerAdded) return;

        Grid dropGrid = this.FindControl<Grid>(nameof(Grid_DropGrid));
        dropGrid?.AddHandler(DragDrop.DropEvent, Grid_DropFiles);
        _isDropHandlerAdded = true;
    }

    private async void Grid_DropFiles(object? sender, DragEventArgs e)
    {
        bool areFilesUploaded = false;
        string[] allowedFileExtensions = { ".mp3", ".wav", ".ogg", ".flac" };

        if (!e.Data.Contains(DataFormats.Files)) return;

        IEnumerable<IStorageItem>? files = e.Data.GetFiles();
        foreach (IStorageItem file in files)
        {
            string fileExtension = file.GetFileInfo().Extension;
            if (allowedFileExtensions.Contains(fileExtension.ToString()))
            {
                areFilesUploaded = _viewModel.CreateNewBeatAndAddAudioFile(file.GetFileInfo());
            } else
            {
                // Show error box.
            }
        }

        if (areFilesUploaded)
        {
            // The data binding doesn't work in the UserControl for some reason, so I am doing this until I find a fix.
            StackPanel_NoUploadsFound.IsVisible = false;
        } else
        {
            // Give the UI the error message here.
        }
    }

```

**Comment by IanRawley** [comment]
What's usually the case with these, especially when the View is creating its own DataContext in the constructor, is that you're ending up with two instances of the ViewModel.  One created by the view and observed by you, and one created externally and assigned to DataContext which is manipulated by the bindings.  You said you checked that the DataContext was set correctly, did you check that it and the _viewModel field were the same instance?

**Reply by stevemonaco** [reply]
You can sometimes troubleshoot this by putting a breakpoint in the ViewModel ctor and ensuring you aren't creating more than expected.

**Reply by mezmeri** [reply]
You're right. It actually creates two instances of the UploadBeatViewModel. One through the MainViewModel, which handles navigation and once again in the UploadBeatView. 

I've tried figuring out a solution for it, but I can't really think of any since DI is not supported (as far as I can see in the discussions?).

**Reply by stevemonaco** [reply, marked as answer]
You should avoid constructors with parameters for Views, IMO. There are some patterns that "work" but bring some downsides. You're obviously creating the VM in a separate location, so you should find out where. I generally use this pattern for root views:

```cs
var someView = new SomeView(); // Or retrieved via locator
someView.DataContext = someViewModel;
```

This is effectively "parameter injection" style of DI. Other controls below the root, like `ContentControl`, will change the `DataContext` appropriately behind-the-scenes. Then rewrite your View like:

```cs
public partial class UploadBeatView : UserControl
{
    public UploadBeatViewModel ViewModel => (UploadBeatViewModel)DataContext!; // Do not access until after construction
    private bool _isDropHandlerAdded = false;

    public UploadBeatView()
    {
        InitializeComponent();
    }
```

You can do any significant work that requires the ViewModel in `Loaded` or `AttachedToVisualTree`.

Alternatively, you can handle `OnDataContextChanged` and cache the externally set VM from there. I find it's better to fail fast as the VM being unavailable after construction is a developer error in my pattern. I've tried caching before and you need to tiptoe around nullability warnings in every code-behind handler which was adding negative value from the extra code required.

**Reply by mezmeri** [reply]
Oh yeah, I can definitely see that now. I have never worked with a ViewLocator before _(ahem)_, but stepping through the code, it makes sense that it is setting the DataContext. After removing the excess VM instantiation the bindings work... which makes sense.

Thank you for taking the time to help out🥇

**Reply by mezmeri** [reply]
Actually, I have a question regarding your code for the root-views.
```cs
var someView = new SomeView(); // Or retrieved via locator
someView.DataContext = someViewModel;
```
Do you instantiate (or set) both the View and the ViewModel in the same class? I'd like to implement it, however, having a ViewModel (in my case MainViewModel) handle views seems like hack. 🤔

**Reply by stevemonaco** [reply]
To be clearer, I refer only to the topmost Views as a root view. ie. Any `Window` including dialogs on a desktop app.

1. The initial VM + View are created in `OnFrameworkInitializationCompleted`. These are created by DI though I only have parameters for VMs (injecting child VMs and services)
2. The initial View is shown. Controls such as `ContentControl` and collection controls have their content binded to child VMs.
3. This binding automatically creates the View behind-the-scenes via ViewLocator as Avalonia tries to figure out what content to display for a VM type. You can almost as easily use `DataTemplate` instead because the default ViewLocator only works for desktop and without NativeAOT.

As such, I don't directly create VMs or Views that represent page sections besides the ones required for the root. If you're displaying data, then you'll still need to create VMs, ie. a collection of `PersonViewModel`, either directly or via factory.

VMs should never have access to a View or any View-related (Avalonia) types. There are occasions to break the rule, but only with a good reason.

You can check https://github.com/stevemonaco/NonogramAvalonia for a meatier project though there's only one root view (no dialogs).

**Reply by mezmeri** [reply]
Thank you for sharing your project. So far I've found a lot of inspiration. I'll definitely have to look more into it and see how you've implemented it. 😀
